Fix cross-device mv by switching to copy

seanmorris opened the pull request:
Based on exploration started by @sejas https://github.com/WordPress/playground-tools/pull/116

## What is this PR doing?

This PR modifies the `installAsset` procedure to use a copy operation rather than a rename.

## What problem is it solving?

Emscripten does not support rename operation across different Filesystems. We need to be able to copy files across such a boundary.

## How is the problem addressed?

The `BasePHP.cp` method has been added to copy files and directories recursively. The `installAsset` procedure has been modified to use a copy operation rather than a rename.

## Testing Instructions

### In this project:

These are supporting changes for other libraries, so we'll need to build the project and use `npm link` to use the local packages with other projects, like `playground-tools`.

You can start by navigating to the root of the project in a terminal window and running the following:

```bash
git checkout sm-cross-device-mv
nvm use 20
npm install
npm run build
```

### Testing @php-wasm/web

Start the dev server:

```bash
npm run dev
```

Navigate to http://localhost:5400/website-server/#{%20%22$schema%22:%20%22https://playground.wordpress.net/blueprint-schema.json%22,%20%22landingPage%22:%20%22/wp-admin/plugins.php%22,%20%22preferredVersions%22:%20{%20%22php%22:%20%228.0%22,%20%22wp%22:%20%22latest%22%20},%20%22steps%22:%20[%20{%20%22step%22:%20%22login%22,%20%22username%22:%20%22admin%22,%20%22password%22:%20%22password%22%20},%20{%20%22step%22:%20%22installPlugin%22,%20%22pluginZipFile%22:%20{%20%22resource%22:%20%22wordpress.org/plugins%22,%20%22slug%22:%20%22classic-editor%22%20},%20%22options%22:%20{%20%22activate%22:%20true%20}%20}%20]%20}

Ensure the `classic-editor` plugin is installed and active.

### Testing @php-wasm/node

Link `@wp-playground/blueprints`

```bash
cd dist/packages/playground/blueprints/
npm link
```

Link `@php-wasm/node`

```bash
cd dist/packages/php-wasm/node/ 
npm link
```


#### In `playground-tools`

These are supporting changes for other libraries, so we'll need to use `playground-tools` to test it.

#### Create `b.json`
```json
{
	"$schema": "https://playground.wordpress.net/blueprint-schema.json",
	"landingPage": "/wp-admin/plugins.php",
	"preferredVersions": {
		"php": "8.0",
		"wp": "latest"
	},
	"steps": [
		{
			"step": "login",
			"username": "admin",
			"password": "password"
		},
		{
			"step": "installPlugin",
			"pluginZipFile": {
				"resource": "wordpress.org/plugins",
				"slug": "notification"
			},
			"options": {
				"activate": true
			}
		}
	]
}
```
Then run (again, in `playground-tools`:

```bash
git checkout sm-preventing-memory-leaks
nvm use 20
npm install
npm link @wp-playground/blueprints @php-wasm/node
node dist/packages/wp-now/cli.js start --blueprint=b.json
```

Navigate to http://localhost:8881/wp-admin/plugins.php

Ensure the `notification` plugin is installed and active.

adamziel commented:
The implicit file removal makes me feel icky, even if that's exactly what unix does. 

Imagine someone "moving" a directory from an actual disk directory that's mounted inside EmscriptenFS into a Playground directory inside MEMFS. This wouldn't actually move the files – it would delete them from the disk without leaving any copy.

In a way, a lack of support for cross-filesystem mv could actually be a feature here. It protects the files on the disk from being accidentally wiped out. 

Another concern I have is I'm not sure how to confirm this works across combinations of windows/linux/mac and different block devices/symlinks/hardlinks/tmpfs etc. I'm pretty confident about primitive operations like cp and rm, but once we start reasoning about more advanced cases such as device type detection then there's a lot of ground to cover. A bad outcome would be not covering all of that and then seeing a large number bug reports a few months later.

Intuitively, this feels like something that should be implemented by Emscripten. And maybe it will be in the upcoming `WASMFS` and we'll get it for free? Some research on that would surely be helpful. On the Playground side, it feels much easier to just give up and document lack of support for "heterogenous" `mv` as a known limitation.

I'd like to explore using the [Compression Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Compression_Streams_API) to unzip files as we download them (vs downloading and then unzipping). If that works, we could just write the zip contents directly into `wp-content/plugins/<plugin name>` without ever going through the `tmp` directory, meaning no cross-filesystem writing would be required.

Also, a recursive `copy(src, dst)` function would be useful and would give us all the benefits without any of the downsides. The developers could then lean on an explicit _copy_ and _remove_ calls. 

There's a way to unblock https://github.com/WordPress/playground-tools/pull/116 with a way smaller impact surface. Adding a private `copy` function in `blueprints/common.ts` would allow the Blueprint steps such as `installPlugin` to explicitly copy and remove data. Then, later on, the larger Compression Streams refactoring could remove that code without ever exposing a new public method or adding more low-level `mv` logic.

Oh, and please let's move the polyfills to another PR so 1) they're not blocked by this dicussion and 2) we can reason about them separately.

cc @dmsnell @eliot-akira @sejas – I wonder what do y'all think here

eliot-akira commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -613,7 +613,51 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
 	@rethrowFileSystemError('Could not move "{path}"')
 	mv(fromPath: string, toPath: string) {
-		this[__private__dont__use].FS.rename(fromPath, toPath);
+		const mv = (fromPath: string, toPath: string) => {
```
A small nitpick: it seems unnecessary to create an internal function `mv` here, which has the same signature as the method and is called immediately after creation with the same arguments.

eliot-akira commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -613,7 +613,51 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+				files
+					.filter((f: string) => !['.', '..'].includes(f))
+					.forEach((f: string) =>
+						mv(fromPath + '/' + f, toPath + '/' + f)
```
I suppose this line could use `this.mv()`, instead of having to create an internal function `mv` which has the same signature as the method.

seanmorris replied:
@eliot-akira I thought the same thing, but this method is overridden in some use cases, and those overrides do not include calls to `super` so that breaks the recursion unless I create another function to recurse on.

eliot-akira replied:
Oh I see, I didn't think about overrides, that makes sense!

eliot-akira commented:
> even if that's exactly what unix does

For curiosity's sake, I looked into how `mv` works.

> To move a file, `mv` ordinarily simply renames it. However, if renaming does not work because the destination’s file system differs, `mv` falls back on copying as if by `cp -a`, then (assuming the copy succeeded) it removes the original.
>
> (From https://www.gnu.org/software/coreutils/manual/html_node/mv-invocation.html)

In the [source code for `mv`](https://github.com/coreutils/coreutils/blob/0ee8b03422b134859ceedc5f52cc8bbcf1031f74/src/mv.c#L131C6-L131C15), it calls `cp` with the option `move_mode`. So the moving logic is mainly in [`copy.c`](https://github.com/coreutils/coreutils/blob/master/src/copy.c). Skimming through it, there are a number of situations it handles when moving a file or directory. For example:

- If source and destination files are the same
- If either is `.` or `..`
- If either is a symlink or hardlink
- If source is a symlink and destination is its referent
- If source is a directory and destination is a file (not allowed), or vice versa (allowed)
- "Create a backup of each destination directory in move mode" (And if source or destination is inside `/tmp`, making sure that the backup doesn't clobber either)
- Something about preventing writing through a "dangling symlink"
- "Ensure we don't try to copy through a symlink that was created by a prior call to this function" (I'm guessing it's about when source directory contains a symlink that includes destination)

I don't know how many of these are relevant to our situation, but I can imagine some edge cases that could accidentally result in data loss during recursive copy and remove. So I agree on this point raised:

> Another concern I have is I'm not sure how to confirm this works across combinations of windows/linux/mac and different block devices/symlinks/hardlinks/tmpfs etc. 

This might be why Emscripten only provides `rename` and not `mv` (which we aliased to `FS.rename`).

- emscripten-core/emscripten#18791

On the other hand, if Emscripten doesn't provide `mv`, and if we provide only partial support on same device/filesystem, then users would be responsible for implementing their own moving logic - which might be even more error prone.

eliot-akira commented:
About the polyfills for `CustomEvent` and `File`, since they're about compatibility with Node.js 18, it might be useful to put them in a common place that's convenient for importing from other packages (so they don't have to reach into each other to import the polyfill).

Currently they're in:

- `File` - `packages/playground/blueprints/src/lib/steps/common.ts`
- `CustomEvent` - `packages/php-wasm/progress/src/lib/custom-event.ts` (this PR)

Maybe they can even be exported in a way that external packages like `wp-now` can use them if needed.

seanmorris commented:
I've updated this PR to use a copy operation rather than a move.

I've separated the polyfills into this draft PR: https://github.com/WordPress/wordpress-playground/pull/865

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+
+		const fromStat = FS.stat(fromPath);
+
+		const fromMode = parseInt(fromStat.mode.toString(8).substring(0, 1));
```
Let's make this more human-readable – here are the thoughts I had when I was reading this:

* What's `mode`?
* What's `8`? 
* Why does it need to be `substring`-ed?

seanmorris replied:
@adamziel That's just a way to tell if a link represents a directory or a file. Is this more readable?

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/base-php.ts#L628

adamziel replied:
Thank you for the update! I'm still a bit confused about what's a mode, what's that flag etc:

```ts
		// Emscripten directories will have the sticky-bit set
		// https://linux.die.net/man/1/chmod
		const fromIsDir = fromStat.mode & 0o40000;
```

Let's be kind to readers of this code and explain the concepts and reasoning in place instead of just pointing them to the linux manual.

seanmorris replied:
https://github.com/WordPress/wordpress-playground/blob/cc576fc972ceaabeba7660aaa39494c95aca2de8/packages/php-wasm/universal/src/lib/base-php.ts#L631-L641

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+
+		const fromMode = parseInt(fromStat.mode.toString(8).substring(0, 1));
+
+		const fromIsDir = fromMode === 4;
```
ditto

seanmorris replied:
Ditto:

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/base-php.ts#L628

adamziel replied:
@seanmorris thanks! As a side note – as I mentioned in another PR, links pointing to the `/files` tab become stale once the HEAD pf this branch changes. To share links that remain valid as the branch changes, you'll have to link to changes in specific commits.

seanmorris replied:
> @seanmorris thanks! As a side note – as I mentioned in another PR, links pointing to the `/files` tab become stale once the HEAD pf this branch changes. To share links that remain valid as the branch changes, you'll have to link to changes in specific commits.

I was wondering how you were formatting that code right in the comments. Thanks for showing me that. I've fixed all the links in this PR.

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
 
+	/** @inheritDoc */
+	@rethrowFileSystemError('Could not copy "{path}"')
+	cp(fromPath: string, toPath: string) {
```
Let's add an options arg with a single `recursive: bool` option to ensure the developers express an explicit intention to copy a directory

seanmorris replied:
https://github.com/WordPress/wordpress-playground/pull/846/files#diff-b33e1e6834fb94ac2cb5338e3ae17c60abf7ac88cb057189a1915531e55f7243R621

adamziel replied:
It seems like you added a boolean arg. Let's use an options arg instead, like I mentioned in that previous comment. It would have just a single option, like `recursive: bool`. So you would call it like `cp(fromPath, toPath, { recursive: true })`. This makes the calls clear, as you don't have to think about what that `true` value means, and it also makes the API easy to extend in the future.

seanmorris replied:
@adamziel Ah, I misunderstood. Done.
https://github.com/WordPress/wordpress-playground/blob/0a3ef1eb1a99569921e8238c43314fea8704a174/packages/php-wasm/universal/src/lib/base-php.ts#L622

https://github.com/WordPress/wordpress-playground/blob/0a3ef1eb1a99569921e8238c43314fea8704a174/packages/php-wasm/universal/src/lib/universal-php.ts#L554-L560

adamziel replied:
Thank you! Note that `cp()` signature makes the `options` arg mandatory – let's make it optional like it is in `rmdir`, otherwise you'll always have to call it with an empty object like `cp(fromPath, toPath, {})`

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+			try {
+				FS.mkdir(toPath);
+			} catch {
+				// Do nothing
```
Why is it okay to fail here?

seanmorris replied:
@adamziel Refactored to make it more readable. It only needs to create the directory if it doesn't already exist.

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/base-php.ts#L651

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+				// Do nothing
+			}
+
+			const files = FS.readdir(fromPath);
```
Why not `this.listFiles()`? It already filters out `.` and `..`.

seanmorris replied:
@adamziel I didn't know that existed. Thank you!

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/base-php.ts#L655

adamziel replied:
Thanks! There's a few more functions available on `this` that the `cp` method here could reuse, like `fileExists`. Identifying and applying them could be a good exercise here.

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+			files
+				.filter((f: string) => !['.', '..'].includes(f))
+				.forEach((f: string) =>
+					this.cp(fromPath + '/' + f, toPath + '/' + f)
```
Tip: You can use `joinPaths` from `@php-wasm/util` to safely join paths.

seanmorris replied:
@adamziel Nice! I always prefer to use that type of thing, since paths arent *really* strings.

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/base-php.ts#L657-L664

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+
+			files
+				.filter((f: string) => !['.', '..'].includes(f))
+				.forEach((f: string) =>
```
Let's avoid single-letter variable names and use full words. Looking around the file and seeing just `f` is quite confusing and requires tracing the code back to the line that says `files` and then keeping that context in mind.

seanmorris replied:
@adamziel Done. https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/base-php.ts#L657-L664

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+					this.cp(fromPath + '/' + f, toPath + '/' + f)
+				);
+
+			FS.rmdir(fromPath);
```
Why not `this.rmdir`? It's defined just a few lines below

seanmorris replied:
@adamziel That's been refactored out.

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,39 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+				);
+
+			FS.rmdir(fromPath);
+		} else {
```
Code style: Let's use a guard here instead of a long if/else – it saves some mental cycles by reducing the amount of nesting and context the reader needs to remember.

```js
		if (!fromIsDir) {
			const file = FS.readFile(fromPath, { encoding: 'binary' });
			FS.writeFile(toPath, file);
		}
```

seanmorris replied:
@adamziel You're right, that's a lot more readable.

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/base-php.ts#L621-L664

adamziel commented on packages/php-wasm/universal/src/lib/universal-php.ts:
```diff
@@ -256,6 +256,15 @@ export interface IsomorphicLocalPHP extends RequestHandler {
+	 * Copies a file or directory in the PHP filesystem to a
+	 * new location.
+	 *
+	 * @param oldPath The path to rename.
```
`The path to copy`

seanmorris replied:
https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/universal-php.ts#L263-L267

adamziel commented on packages/php-wasm/universal/src/lib/universal-php.ts:
```diff
@@ -256,6 +256,15 @@ export interface IsomorphicLocalPHP extends RequestHandler {
+	 * new location.
+	 *
+	 * @param oldPath The path to rename.
+	 * @param newPath The new path.
```
Is this a path to the parent directory? Or does it have to include a file name?

seanmorris replied:
@adamziel You need to provide the full, new path. I made it more explicit:

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/universal/src/lib/universal-php.ts#L263-L267

adamziel commented:
It seems like the end to end tests fail on this PR – any ideas about the causes of those failures @seanmorris? Also, let's add some unit test coverage to this change.

seanmorris commented:
> It seems like the end to end tests fail on this PR – any ideas about the causes of those failures @seanmorris? Also, let's add some unit test coverage to this change.

~~Are you referring to the `timeout` errors that happen when we run `npm run test`? If so, I think those might be due to heavy resource usage during the test. It might be worthwhile to explore setting up something to limit parallelism in some cases there. I don't think that failure is totally deterministic.~~

I'll get right on that unit test.

Edit: I extended the timeout on the networking test to 5sec, from 1sec. Looks like that's all it needs to pass on decent hardware.

The copy function also has tests now.

adamziel commented on packages/php-wasm/node/src/test/php-networking.spec.ts:
```diff
@@ -21,7 +21,7 @@ describe.each(SupportedPHPVersions)(
 		});
 	},
-	1000
+	5000
```
Why change this value to 5000?

seanmorris replied:
@adamziel This change was originally related to the e2e crash, which I suspected to be resource related, but I quickly determined that this wasn't the cause once I was able to get things running locally, so I've reverted this change.

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,53 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+		// https://linux.die.net/man/1/chmod
+		const fromIsDir = fromStat.mode & 0o40000;
+
+		let toExists: boolean;
```
I wonder if this would be more readable, what do you think?

```suggestion
		let destinationExists: boolean;
```

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,53 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+		}
+
+		if (!fromIsDir) {
+			const file = FS.readFile(fromPath, { encoding: 'binary' });
```
Isn't `{ encoding: 'binary' }` the default? Also, let's use functions available on `this` whenever possible, e.g. there's `this.readFileAsBuffer()`

seanmorris replied:
https://github.com/WordPress/wordpress-playground/blob/cc576fc972ceaabeba7660aaa39494c95aca2de8/packages/php-wasm/universal/src/lib/base-php.ts#L656

adamziel replied:
Thanks! What about other places where `this` functions could be applicable over custom-built logic?

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +616,53 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+
+		if (!fromIsDir) {
+			const file = FS.readFile(fromPath, { encoding: 'binary' });
+			FS.writeFile(toPath, file);
```
What should happen if `toPath` exists and is a directory?

seanmorris replied:
@adam In that case, `it('cp() should copy a file into a directory', () => {...`

https://github.com/WordPress/wordpress-playground/blob/8da67082fc0347b6a1d7c94c6f1506be3316666c/packages/php-wasm/universal/src/lib/base-php.ts#L658-L665

https://github.com/WordPress/wordpress-playground/blob/cc576fc972ceaabeba7660aaa39494c95aca2de8/packages/php-wasm/node/src/test/php.spec.ts#L351-L360

adamziel replied:
That test only covers a case when `oldPath` is a file, right? There's also a scenario when it's a directory and that also needs a test.

adamziel commented on packages/php-wasm/universal/src/lib/universal-php.ts:
```diff
@@ -256,6 +256,15 @@ export interface IsomorphicLocalPHP extends RequestHandler {
 	mv(oldPath: string, newPath: string): void;
 
+	/**
+	 * Copies a file or directory in the PHP filesystem to a
+	 * new location.
```
Let's thoroughly explain what this function does in different situations, like when the target path already exists.

seanmorris replied:
https://github.com/WordPress/wordpress-playground/blob/cc576fc972ceaabeba7660aaa39494c95aca2de8/packages/php-wasm/universal/src/lib/universal-php.ts#L259-L275

adamziel replied:
Thank you! As I'm read it, I have additional questions:

> The target directory's parent must be a  valid location

A `valid location` means only that it must exist, right? If yes, let's say it directly.

>  If the target path is a file that already exists, it will be overwritten. 

Is that the case only when the source path is a file? Or also when it's a directory?

> If the target path is a directory that already exists, the file or directory will be copied into it.

So if the target path is a directory that does not exist yet, will it be created?

adamziel replied:
I also wonder about the semantics here. Right now this function makes an implicit assumption that says:

> If the `newPath` is a directory that already exists, the developer intended to copy the `oldPath` inside that directory.

I don't think it always holds. Maybe the `newPath` shouldn't actually be there? I'd rather remove the implicit assumption and make the logic explicit: If `newPath` is a directory that already exists, that's an error, please remove it explicitly before the recursive copy.

adamziel commented on packages/php-wasm/web/src/lib/web-php-endpoint.ts:
```diff
@@ -89,6 +89,11 @@ export class WebPHPEndpoint implements IsomorphicLocalPHP {
 		return _private.get(this)!.php.mv(fromPath, toPath);
 	}
 
+	/** @inheritDoc @php-wasm/universal!IsomorphicLocalPHP.mv  */
```
Should this be updated?

seanmorris replied:
https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/packages/php-wasm/web/src/lib/web-php-endpoint.ts#L92-L95

adamziel commented on packages/playground/website/cypress/e2e/app.cy.ts:
```diff
@@ -104,7 +104,7 @@ describe('Query API', () => {
 				.find('[data-slug=gutenberg].active')
 				.should('exist');
-		});
+		}, 120_000);
```
What's 120_000?

seanmorris replied:
@adamziel the `_` is an ignored character when it appears directly between two digits. It's just a thousands separator.

This change was originally related to the e2e crash, which I suspected to be resource related, but I quickly determined that this wasn't the cause once I was able to get things running locally, so I've reverted this change.

seanmorris commented:
The following error was a red herring. The failure was happening directly before it, but the crash doesn't seem to happen once the **actual** bug has been fixed.

```
We detected that the Chromium Renderer process just crashed.

This is the equivalent to seeing the 'sad face' when Chrome dies.

This can happen for a number of different reasons:
```

I wasn't able to run the e2e tests locally until I read a bunch of docs and added the following lines:

https://github.com/WordPress/wordpress-playground/blob/cd3902f0672dd2d3dcc7eb9762ed816fcd9c6267/nx.json#L12-L14
 
Once I added that I was able to run `nx affected --target=e2e --configuration=ci --verbose` and I determined the error was coming from `nx run playground-website:e2e`.

Once I zeroed in on that I was able to get the e2e tests passing again.

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -616,6 +617,77 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
+			destinationExists = false;
+		}
+
+		if (!sourceIsDir) {
```
What will happen if I copy a directory tree like `wordpress/wp-content/plugins` onto another directory, say `/www` that already has `wordpress/wp-content/plugins` in it? It seems like it would duplicate some directories?

adamziel commented:
> I wasn't able to run the e2e tests locally until I read a bunch of docs and added the following lines:

Aha! What was the error? Do you mean the Chrome Crash, or was there another issue? If that was the chrome crash, I wonder why that config update would help 🤔  If that fix be useful for others, you could start an issue to share it with others by updating the docs or the repo setup.
